File: macos/Sources/JasnostCaptureCore/ActivityEvent.swift
import Foundation

/// Bounding box of a UI element, in screen points.
public struct BoundingBox: Codable, Sendable, Equatable {
    public var x: Double
    public var y: Double
    public var width: Double
    public var height: Double

    public init(x: Double, y: Double, width: Double, height: Double) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

/// A point in screen coordinates — the release point of a drag (``ActivityEvent.dragEnd``).
public struct DragPoint: Codable, Sendable, Equatable {
    public var x: Double
    public var y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

/// Namespaced OS application identity. This is deliberately separate from document/page URL and
/// from a later derived business-system identity.
public struct ActivityApplicationIdentity: Codable, Sendable, Equatable {
    public var namespace: String
    public var value: String
    public var name: String?
    public var version: String?

    public init(namespace: String, value: String, name: String? = nil, version: String? = nil) {
        self.namespace = namespace
        self.value = value
        self.name = name
        self.version = version
    }
}

/// The semantic target of an action — the desktop equivalent of a DOM element. Built from
/// the Accessibility (AX) element hit-tested under the cursor. Optional fields are omitted
/// from JSON when nil (Swift synthesises `encodeIfPresent` for optionals).
public struct EventTarget: Codable, Sendable, Equatable {
    public var tag: String?
    public var role: String?
    public var accessibleName: String?
    public var text: String?
    public var boundingBox: BoundingBox?

    public init(
        tag: String? = nil,
        role: String? = nil,
        accessibleName: String? = nil,
        text: String? = nil,
        boundingBox: BoundingBox? = nil
    ) {
        self.tag = tag
        self.role = role
        self.accessibleName = accessibleName
        self.text = text
        self.boundingBox = boundingBox
    }
}

/// One captured activity event, matching `contract/schema/activity-event.schema.json`
/// (the same contract the Chrome extension emits). The schema is
/// `additionalProperties: false`, so only these fields may appear; nil optionals are
/// omitted. `sessionId`, `eventId`, `timestamp`, `eventType`, `url` are required.
public struct ActivityEvent: Codable, Sendable, Equatable {
    public var sessionId: String
    public var eventId: String
    public var sequence: Int?
    public var timestamp: String
    public var eventType: String
    /// Legacy context carrier retained for compatibility. Native clients keep application identity
    /// here as `app://<bundle-id>` and put sanitized page/document context in `documentURL`.
    public var url: String
    public var application: ActivityApplicationIdentity?
    /// Sanitized observed document/page URL. It is context evidence, never application or stable
    /// business-object identity.
    public var documentURL: String?
    public var pageTitle: String?
    public var system: String?
    public var target: EventTarget?
    public var value: String?
    /// Text the user had selected at the interaction (AX kAXSelectedText) — e.g. a word
    /// double-clicked or a range drag-selected. The SELECTION, distinct from `target.text` (the
    /// element's full value). Masked when the field is sensitive.
    public var selectedText: String?
    /// Bounded clipboard payload observed for paste. Copy/cut use `selectedText`, because the
    /// pasteboard may still be stale at key-down. Evidence only; not proof of a successful transfer.
    public var clipboardText: String?
    /// OS click state for a click/contextmenu/drag: 1 = single, 2 = double, 3 = triple.
    public var clickCount: Int?
    /// For a `drag` event: the point where the mouse button was released (start = the event location).
    public var dragEnd: DragPoint?
    /// One locally-minted identifier for the physical gesture. A drag is emitted once, on mouse-up;
    /// the ID supports correlation without forcing downstream gesture inference.
    public var gestureId: String?
    public var inputMasked: Bool?
    public var isSensitive: Bool?
    public var screenshotId: String?
    public var audioFileId: String?
    /// Stable id of the active label (bracketed segment), minted at label start. Stamped onto
    /// every event captured while a label is open; nil when no label is active. Set on
    /// `label_start`/`label_end` events, which are the segment boundaries themselves.
    public var labelId: String?
    /// Human-readable name of the active label; nil when no label is active.
    public var label: String?
    /// Stable id of the Process (from the Area's inventory) this bracketed segment demonstrates —
    /// stamped per label, like `labelId` (the capture-side wiring lands with the Guided picker, a
    /// later phase; this carrier just defines the field). nil when the segment isn't process-anchored.
    /// Segment-scoped; the session's Area rides the OTLP attribute `area.id`, not an event field. ADR 0002.
    public var processId: String?
    /// Human-readable Process name (e.g. "Invoicing"); nil when the segment isn't process-anchored.
    public var process: String?
    /// Local-only: an inline `data:image/...;base64,…` screenshot. The agent uploads the image
    /// to Keboola Files and replaces it with `screenshotId` before emitting the event, so this
    /// field never goes on the wire; raw image bytes never reach OTLP — only the `screenshotId`
    /// reference does.
    public var screenshotDataUrl: String?

    public init(
        sessionId: String,
        eventId: String,
        sequence: Int? = nil,
        timestamp: String,
        eventType: String,
        url: String,
        application: ActivityApplicationIdentity? = nil,
        documentURL: String? = nil,
        pageTitle: String? = nil,
        system: String? = nil,
        target: EventTarget? = nil,
        value: String? = nil,
        selectedText: String? = nil,
        clipboardText: String? = nil,
        clickCount: Int? = nil,
        dragEnd: DragPoint? = nil,
        gestureId: String? = nil,
        inputMasked: Bool? = nil,
        isSensitive: Bool? = nil,
        screenshotId: String? = nil,
        audioFileId: String? = nil,
        labelId: String? = nil,
        label: String? = nil,
        processId: String? = nil,
        process: String? = nil,
        screenshotDataUrl: String? = nil
    ) {
        self.sessionId = sessionId
        self.eventId = eventId
        self.sequence = sequence
        self.timestamp = timestamp
        self.eventType = eventType
        self.url = url
        self.application = application
        self.documentURL = documentURL
        self.pageTitle = pageTitle
        self.system = system
        self.target = target
        self.value = value
        self.selectedText = selectedText
        self.clipboardText = clipboardText
        self.clickCount = clickCount
        self.dragEnd = dragEnd
        self.gestureId = gestureId
        self.inputMasked = inputMasked
        self.isSensitive = isSensitive
        self.screenshotId = screenshotId
        self.audioFileId = audioFileId
        self.labelId = labelId
        self.label = label
        self.processId = processId
        self.process = process
        self.screenshotDataUrl = screenshotDataUrl
    }
}

/// Event types the agent emits (subset of the schema enum relevant to desktop capture).
public enum EventType: String, Sendable {
    case sessionStart = "session_start"
    case sessionEnd = "session_end"
    case navigate  // app/window switch
    case click
    case contextmenu
    case input
    case change
    case copy
    case cut
    case paste
    case scroll
    case drag  // mouse drag (drag-select / drag-and-drop): start = location, end = dragEnd
    case keydown
    case focus
    case screenshot
    case narration
    case annotation  // user-declared task label (live labeling) — opens a labeled segment
    case labelStart = "label_start"  // opens a bracketed label segment
    case labelEnd = "label_end"  // closes the open bracketed label segment
}

File: windows/Sources/JazzCaptureCore/ActivityEvent.cs
using System.Text.Json.Serialization;

namespace JazzCaptureCore;

/// <summary>
/// The application a captured event happened in.
/// </summary>
/// <param name="Namespace">Identifier namespace, e.g. <c>windows.aumid</c> or <c>macos.bundle-id</c>.</param>
/// <param name="Value">The identifier itself. Projected as <c>application.id</c> — the key is renamed.</param>
/// <param name="Name">Human-readable application name.</param>
/// <param name="Version">Application version string.</param>
public sealed record ApplicationRef(
    [property: JsonPropertyName("namespace")] string Namespace,
    [property: JsonPropertyName("value")] string Value,
    [property: JsonPropertyName("name")] string? Name = null,
    [property: JsonPropertyName("version")] string? Version = null);

/// <summary>Element rectangle in screen coordinates. All four components are required together.</summary>
public sealed record BoundingBox(
    [property: JsonPropertyName("x")] double X,
    [property: JsonPropertyName("y")] double Y,
    [property: JsonPropertyName("width")] double Width,
    [property: JsonPropertyName("height")] double Height);

/// <summary>Release point of a completed drag gesture.</summary>
public sealed record DragEnd(
    [property: JsonPropertyName("x")] double X,
    [property: JsonPropertyName("y")] double Y);

/// <summary>
/// The UI element an event was directed at, as resolved by the accessibility layer.
/// </summary>
/// <remarks>
/// <see cref="Text"/> is retained for archive records but is <em>never</em> projected to OTLP —
/// it can hold user content. The schema's <c>selectorCandidates</c> and <c>attributes</c> members
/// are browser-only and are ignored on input; the projection emits a constant empty
/// <c>target.selectorCandidates</c> instead.
/// </remarks>
public sealed record EventTarget
{
    [JsonPropertyName("tag")]
    public string? Tag { get; init; }

    [JsonPropertyName("role")]
    public string? Role { get; init; }

    [JsonPropertyName("accessibleName")]
    public string? AccessibleName { get; init; }

    [JsonPropertyName("text")]
    public string? Text { get; init; }

    [JsonPropertyName("boundingBox")]
    public BoundingBox? BoundingBox { get; init; }
}

/// <summary>
/// One captured activity event, mirroring <c>contract/schema/activity-event.schema.json</c>.
/// </summary>
/// <remarks>
/// Unknown JSON properties are ignored on purpose: browser-only members (<c>tabId</c>,
/// <c>frameId</c>, <c>viewport</c>, <c>rrwebChunkId</c>) travel through the same schema but have
/// no desktop meaning, and tolerating them keeps the model forward-compatible with schema
/// additions. Absent optional members stay <see langword="null"/> — the OTLP projection, not the
/// model, decides whether a missing value becomes an empty string or an omitted key.
/// </remarks>
public sealed record ActivityEvent
{
    [JsonPropertyName("sessionId")]
    public string SessionId { get; init; } = string.Empty;

    [JsonPropertyName("eventId")]
    public string EventId { get; init; } = string.Empty;

    /// <summary>Monotonic per-session ordinal. Numeric: omitted from OTLP when absent.</summary>
    [JsonPropertyName("sequence")]
    public int? Sequence { get; init; }

    /// <summary>ISO 8601 capture instant; the source of the record's nanosecond timestamps.</summary>
    [JsonPropertyName("timestamp")]
    public string Timestamp { get; init; } = string.Empty;

    /// <summary>One of the 24 contract event types. Only <c>narration</c> changes the projection.</summary>
    [JsonPropertyName("eventType")]
    public string EventType { get; init; } = string.Empty;

    [JsonPropertyName("url")]
    public string Url { get; init; } = string.Empty;

    [JsonPropertyName("application")]
    public ApplicationRef? Application { get; init; }

    /// <summary>Projected as <c>document.url</c>.</summary>
    [JsonPropertyName("documentURL")]
    public string? DocumentUrl { get; init; }

    [JsonPropertyName("pageTitle")]
    public string? PageTitle { get; init; }

    /// <summary>Legacy display hint (the owning application's short name).</summary>
    [JsonPropertyName("system")]
    public string? System { get; init; }

    [JsonPropertyName("target")]
    public EventTarget? Target { get; init; }

    /// <summary>Typed value, already redacted by the capture host.</summary>
    [JsonPropertyName("value")]
    public string? Value { get; init; }

    [JsonPropertyName("selectedText")]
    public string? SelectedText { get; init; }

    [JsonPropertyName("clipboardText")]
    public string? ClipboardText { get; init; }

    /// <summary>OS click multiplicity (1/2/3). Numeric: omitted from OTLP when absent.</summary>
    [JsonPropertyName("clickCount")]
    public int? ClickCount { get; init; }

    /// <summary>Drag release point. Numeric group: both keys omitted together when absent.</summary>
    [JsonPropertyName("dragEnd")]
    public DragEnd? DragEnd { get; init; }

    [JsonPropertyName("gestureId")]
    public string? GestureId { get; init; }

    /// <summary>True when <see cref="Value"/> was masked. Defaults to <see langword="false"/> in OTLP.</summary>
    [JsonPropertyName("inputMasked")]
    public bool? InputMasked { get; init; }

    [JsonPropertyName("isSensitive")]
    public bool? IsSensitive { get; init; }

    [JsonPropertyName("screenshotId")]
    public string? ScreenshotId { get; init; }

    /// <summary>Audio blob reference; only meaningful on <c>narration</c> events.</summary>
    [JsonPropertyName("audioFileId")]
    public string? AudioFileId { get; init; }

    [JsonPropertyName("labelId")]
    public string? LabelId { get; init; }

    /// <summary>Projected as <c>label.name</c>.</summary>
    [JsonPropertyName("label")]
    public string? Label { get; init; }

    [JsonPropertyName("processId")]
    public string? ProcessId { get; init; }

    /// <summary>Projected as <c>process.name</c>.</summary>
    [JsonPropertyName("process")]
    public string? Process { get; init; }
}
